File: solutions/1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.rb
def max_equal_rows_after_flips(mat)
    pat_freq = Hash.new(0)
    
    mat.each do |row|
        pattern = row[0] == 0 ? 
            row.join : 
            row.map { |bit| bit ^ 1 }.join
            
        pat_freq[pattern] += 1
    end
    
    pat_freq.values.max
end

File: solutions/1147-flip-columns-for-maximum-number-of-equal-rows/flip-columns-for-maximum-number-of-equal-rows.py
class Solution:
    def maxEqualRowsAfterFlips(self, matrix: List[List[int]]) -> int:
        pattern_frequency = Counter()
        
        for current_row in matrix:
            canonical_pattern = tuple(current_row) if current_row[0] == 0 else tuple(bit ^ 1 for bit in current_row)
            pattern_frequency[canonical_pattern] += 1
            
        return max(pattern_frequency.values())